cpp/src/io/utilities/hostdevice_vector.hpp

```diff
@@ -124,20 +124,42 @@ class hostdevice_vector {
   operator cudf::host_span<T>() { return {h_data, max_elements}; }
   operator cudf::host_span<T const>() const { return {h_data, max_elements}; }
 
-  void host_to_device(rmm::cuda_stream_view stream, bool synchronize = false)
+  void host_to_device(size_t start_pos,
```

Comment:
I'm not sure how I feel about the partial copy methods here. At first, I was thinking we would solve this problem by using a different container (e.g. `std::vector` on the host) and use [vector factories](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/include/cudf/detail/utilities/vector_factories.hpp) with the spans of items to be copied (to an `rmm::uvector` on the device). I would prefer that approach since it seems the utility of subspan copying on the `hostdevice_vector` class is limited. I realize I didn't communicate very clearly, but this is what I had in mind when I [said](https://github.com/rapidsai/cudf/pull/10699#discussion_r858162650) "If this only copies a subset of the elements, it seems unintuitive to use a `hostdevice_vector` [...] Maybe a correctly-sized container that automatically converts to `device_span` would be better."

`hostdevice_vector` seems most appropriate when we know we want pinned memory that can copy efficiently back and forth many times. If I recall correctly, the host/device copies in this case are not happening many times in a row, so maybe having the pinned host/device container isn't really needed. Vector factories that copy a subspan of memory might be equally good for this atypical use case of copying subspans.

Also, adding these methods opens up `hostdevice_vector` to a new class of out-of-range errors where `start_pos + count > size()`. I would not want to add bounds checking to a container primitive, so documentation listing the requirements would be important if we choose to go this path.

What do you think about reverting this and using vector factories to copy host/device data from the appropriate spans? I think it keeps the scope of this PR tighter and avoids introducing single-use behavior.

Reply:
Aside from this issue, I'm ready to approve the PR. Happy to discuss pros/cons with you before implementing changes.